{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Quick and clean: Python for biological data processing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- [Course information](day1/info.ipynb)\n",
    "    - About the course\n",
    "    - About us\n",
    "    - About you\n",
    "    - Venue, credits, pricing, requirements\n",
    "    - Feedback\n",
    "    - Bibliography\n",
    "\n",
    "## Day 1, 8:00-16:00, Python\n",
    "\n",
    "Easy introduction, light tasks.\n",
    "\n",
    "- [What is Python?](day1/what.ipynb)\n",
    "    - How do computers work?\n",
    "    - Python: Stats, strengths and weaknesses\n",
    "    - Python: Past, present and future\n",
    "    - As a new \"pythonista\", who will be your friends?\n",
    "    - Make Python work again!\n",
    "    - Installing libraries\n",
    "    - Python consoles, interpreters and editors\n",
    "    - Python distributions, conda\n",
    "\n",
    "\n",
    "- [Jupyter](day1/jupyter.ipynb)\n",
    "    - History and JupyterLab\n",
    "    - General presentation\n",
    "    - Use in reproducible research\n",
    "    - Other uses\n",
    "    - Slide shows\n",
    "    \n",
    "    \n",
    "- [Python tutorial](day1/tutorial.ipynb)\n",
    "    - Basics: Math, Variables, Functions, Control flow, Modules\n",
    "    - Data representation: String, Tuple, List, Set, Dictionary, Objects and Classes\n",
    "    - Standard library modules: script arguments, file operations, timing, processes, forks, multiprocessing\n",
    "    \n",
    "    \n",
    "- [DevOps](day1/devops.ipynb)\n",
    "    - Reproducible research.\n",
    "    - What are devops?\n",
    "    - How to source your code.\n",
    "    - Using source editors. What matters?\n",
    "    - Distributed version control using git.\n",
    "    - Development vs production.\n",
    "    - When do we need containers? Using Docker.\n",
    "    - Speed: Profiling, IPython, JIT \n",
    "    - Robustness: unit testing\n",
    "    - Documentation: pydoc and Sphinx\n",
    "    \n",
    "    \n",
    "- [Python and the data](day1/web.ipynb)\n",
    "    - File IO, streaming, serialization\n",
    "    - Parsing and regular expressions\n",
    "    - Chunking and HDF5, pytables\n",
    "    - XML, HTML editing\n",
    "    - Web frameworks.\n",
    "    - SQL frameworks.\n",
    "    - Remote API calls (Entrez, BioBank)\n",
    "    \n",
    "    \n",
    "- [Python and other languages](day1/languages.ipynb)\n",
    "    - Interacting with other programs through API calls.\n",
    "    - The whole is greater than its parts: syncretism.\n",
    "    - Python and C: make your own API.\n",
    "    - Python and R: direct piping through Jupyter.\n",
    "    - Python and Julia\n",
    "    \n",
    "    \n",
    "- [Python and the cloud](day1/cloud.ipynb)\n",
    "    - What is the cloud?\n",
    "    - Hadoop and Spark: distributed computing intro.\n",
    "    - HPC: grid computing intro.\n",
    "    - Spinning instances: Back to containers.\n",
    "    - Accelerate your code: GPU computing, FPGA, ASICS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Day 2, 8:00-16:00, Data science\n",
    "\n",
    "Intensive in math, slightly harder tasks to accomplish in class.\n",
    "\n",
    "- [Intro to data science](day2/introds.ipynb)\n",
    "    - You are a data scientist.\n",
    "    - How to extract information from data.\n",
    "    - Dataset, model, prediction.\n",
    "    \n",
    "    \n",
    "- [Visualization](day2/visualization.ipynb):\n",
    "    - Standard plots with matplotlib and seaborn: line, scatter, chart\n",
    "    - Web publishing with plotly and bokeh: heatmap example\n",
    "    - Network display with graphviz\n",
    "    - GUI programming with wxpython\n",
    "    - Web interfaces\n",
    "    \n",
    "    \n",
    "- [Statistics](day2/statistics.ipynb):\n",
    "    - Dataframing with pandas\n",
    "    - Normalizing a dataset\n",
    "    - Scipy: ANOVA\n",
    "    - Statistical enrichment analysis\n",
    "    \n",
    "    \n",
    "- [Scientific computing](day2/scicomp.ipynb):\n",
    "    - Numpy: advanced array operations\n",
    "    - Scipy introduction: from linear algebra to image analisys\n",
    "    - Simpy: symbolic math\n",
    "    - Networks with networkx: centrality computation\n",
    "    - Fitting a curve, cleaning a signal.\n",
    "    \n",
    "    \n",
    "- [Machine learning](day2/learning.ipynb):\n",
    "    - scikit-learn: clustering\n",
    "    - Handling multivariate data: PCA, PLS regression\n",
    "    - Deep learning: theano, tensorflow and pytorch\n",
    "    \n",
    "    \n",
    "- [Workflow management](day2/workflow.ipynb)\n",
    "    - Snakemake tutorial\n",
    "    - Nextflow tutorial\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Day 3, 8:00-16:00, 'Omics\n",
    "\n",
    "I setup the problems and describe the tasks, and give you some helper code to start with, and you will work on your picked task in class. I will tend to guide rather than tell.\n",
    "\n",
    "- [NGS pipelining](day3/ngs_pipelining.ipynb):\n",
    "    - Open a cloud instance and install required programs\n",
    "    - Setup the pipeline\n",
    "    - Read mapping and IGV inspection\n",
    "    - Normalizing counts and differential expression\n",
    "    - Galaxy integration\n",
    "    \n",
    "    \n",
    "- [Sequencing](day3/sequencing.ipynb)\n",
    "    - Make a toy sequencing library in standard Python for processing DNA, RNA and protein data.\n",
    "    - Implement the DNA, RNA and proteins as Python classes\n",
    "    - Make methods for transcription, translation, regulation.\n",
    "    - Compute several sequence similarity scores, such as hamming distance and mutual information.\n",
    "    - Add BioPython methods and prefix them with bp\n",
    "    - Describe your module in a tutorial like fashion\n",
    "    \n",
    "    \n",
    "- [Gene Expression](day3/expression.ipynb)\n",
    "    - Download a GEO dataset and prepare it\n",
    "    - Cluster the genes based on their expression\n",
    "    - Compute a co-expression network\n",
    "    - Compute differential gene expression for a set of samples.\n",
    "    - Compute functional enrichment of the main clusters.\n",
    "    \n",
    "    \n",
    "- [Transcriptomics](day3/transcriptomics.ipynb)\n",
    "    - Extract the promoter regions using biopython\n",
    "    - Investigate de-novo motifs on clusters of genes using meme or steme\n",
    "    - Use TransFac database to search for motif occurences on selected genes\n",
    "    - Save the found motifs and related data\n",
    "    - Test which of the motif occurences on your selected genes is significant\n",
    "    \n",
    "    \n",
    "- [Proteomics](day3/proteomics.ipynb)\n",
    "    - Compute a protein similarity graph, cluster enrichment study\n",
    "    - Perform structural alignment and plots with PyMol\n",
    "    \n",
    "    \n",
    "- [Metabolomics](day3/metabolomics.ipynb)\n",
    "    - Metabolic pathway assembly, enrichment and display\n",
    "    - Flux balance analysis\n",
    "    \n",
    "    \n",
    "- [Dynamic modeling](day3/dynamic_modeling.ipynb)\n",
    "    - Load a curated SBML model\n",
    "    - Plot the model\n",
    "    - Solve the model\n",
    "    - Peak identification\n",
    "    - Pathway studies\n",
    "    \n",
    "    \n",
    "- [Population genetics and philogeny](PopGen.ipynb) - Removed/Not included.\n",
    "    - Run a small scale coalescent simulation\n",
    "    - Compute a philogeny tree and display it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
