{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sequencing\n",
    "\n",
    "- Make a toy sequencing library in standard Python for processing DNA, RNA and protein data.\n",
    "- Implement the DNA, RNA and proteins as Python classes\n",
    "- Make methods for transcription, translation, regulation.\n",
    "- Compute several sequence similarity scores, such as hamming distance and mutual information, \n",
    "\n",
    "This task has less to do with NGS procedures and more to do with the processing of raw sequences and of the main files involved in sequencing. It is also a bit more scholarly in that, while it is good to use certain specialized libraries, each subtask should also be performed without them. Use the template supplied bellow."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class BaseSequence(object):\n",
    "    \n",
    "    def __init__(self, seqstring):\n",
    "        self.seq = seqstring\n",
    "        return\n",
    "    \n",
    "    def cleanup_sequence(self, seqstring):\n",
    "        \"\"\"\n",
    "        Use self.ValidChars to get the valid characters for each class.\n",
    "        \"\"\"\n",
    "        cleanedseq = \"\"\n",
    "        return cleanedseq\n",
    "    \n",
    "    def append_sequence(self, seqstring):\n",
    "        \"\"\"\n",
    "        Append to self.seq. Validate the sequence first!\n",
    "        \"\"\"\n",
    "        return\n",
    "\n",
    "    def cg_content(self):\n",
    "        \"\"\"\"Return the percentage of G and C characters in the sequence\"\"\"\n",
    "        return\n",
    "    \n",
    "    def get_ORFs(self):\n",
    "        \"\"\"\n",
    "        Return all possible open reading frames\n",
    "        \"\"\"\n",
    "        orfs = set()\n",
    "        return orfs\n",
    "    \n",
    "    def load_from_fasta(self, fname):\n",
    "        \"\"\"Aside from loading the sequence, a few stats should also be displayed\"\"\"\n",
    "        return\n",
    "\n",
    "class DNASequence(BaseSequence):\n",
    "    ValidChars = 'TCAG'\n",
    "    \n",
    "    def __init__(self, seqstring):\n",
    "        super(DNASequence, self).__init__(seqstring)\n",
    "        #super(DNASequence, self).__init__(seqstring)\n",
    "        return\n",
    "\n",
    "    def transcribe(self):\n",
    "        \"\"\"Task 1.2\n",
    "        Produce an RNASequence string. You have an example, but as an \n",
    "        exercise make a more explicit code doing the same thing.\n",
    "        \"\"\"\n",
    "        transeq = \"\"        \n",
    "        #import string\n",
    "        #tbl = string.maketrans('TCGA', 'AGCU')\n",
    "        #transeq = self.seq.translate(tbl)\n",
    "        rna = RNASequence(transeq)\n",
    "        return rna\n",
    "    \n",
    "    def read_fasta(self, fpath):\n",
    "        \"\"\"Task 1.4\n",
    "        Update self.seq\n",
    "        \"\"\"\n",
    "        return\n",
    "    \n",
    "    def binding_site_pos(self, pattern):\n",
    "        \"\"\"\n",
    "        Ex: CCAXXXTGXXXCGG. Use regex.\n",
    "        \"\"\"\n",
    "        pos = -1\n",
    "        return pos\n",
    "    \n",
    "    def restriction_site_pos2(self, setp):\n",
    "        \"\"\"\n",
    "        setp - set of patterns\n",
    "        Example: CCGCGG, CCGTGG, CCACGG, or CCATGG\n",
    "        The purpose of this method is to find establish a regex pattern automatically\n",
    "        based on the given sequences and then apply it\n",
    "        \"\"\"\n",
    "        pos = -1\n",
    "        return pos\n",
    "        \n",
    "    def get_correct_ORF(self):\n",
    "        \"\"\"Task 2.6\n",
    "        Use the parent class getORFs method.\n",
    "        An open reading frame starts with the start codon ATG (Met) \n",
    "        in most species and ends with one of the three stop codons (TAA, TAG or TGA)\n",
    "        \"\"\"\n",
    "        return\n",
    "    \n",
    "    def direct_translation(self):\n",
    "        \"\"\"Task 2.7\n",
    "        Use the RNASequence.pmap\n",
    "        \"\"\"\n",
    "        pseq = \"\"\n",
    "        prot = ProteinSequence(pseq)\n",
    "        return prot\n",
    "        \n",
    "\n",
    "class RNASequence(BaseSequence):\n",
    "    ValidChars = 'AGCU'\n",
    "    pmap = {\"UUU\":\"F\", \"UUC\":\"F\", \"UUA\":\"L\", \"UUG\":\"L\",\n",
    "    \"UCU\":\"S\", \"UCC\":\"S\", \"UCA\":\"S\", \"UCG\":\"S\",\n",
    "    \"UAU\":\"Y\", \"UAC\":\"Y\", \"UAA\":\"STOP\", \"UAG\":\"STOP\",\n",
    "    \"UGU\":\"C\", \"UGC\":\"C\", \"UGA\":\"STOP\", \"UGG\":\"W\",\n",
    "    \"CUU\":\"L\", \"CUC\":\"L\", \"CUA\":\"L\", \"CUG\":\"L\",\n",
    "    \"CCU\":\"P\", \"CCC\":\"P\", \"CCA\":\"P\", \"CCG\":\"P\",\n",
    "    \"CAU\":\"H\", \"CAC\":\"H\", \"CAA\":\"Q\", \"CAG\":\"Q\",\n",
    "    \"CGU\":\"R\", \"CGC\":\"R\", \"CGA\":\"R\", \"CGG\":\"R\",\n",
    "    \"AUU\":\"I\", \"AUC\":\"I\", \"AUA\":\"I\", \"AUG\":\"M\",\n",
    "    \"ACU\":\"T\", \"ACC\":\"T\", \"ACA\":\"T\", \"ACG\":\"T\",\n",
    "    \"AAU\":\"N\", \"AAC\":\"N\", \"AAA\":\"K\", \"AAG\":\"K\",\n",
    "    \"AGU\":\"S\", \"AGC\":\"S\", \"AGA\":\"R\", \"AGG\":\"R\",\n",
    "    \"GUU\":\"V\", \"GUC\":\"V\", \"GUA\":\"V\", \"GUG\":\"V\",\n",
    "    \"GCU\":\"A\", \"GCC\":\"A\", \"GCA\":\"A\", \"GCG\":\"A\",\n",
    "    \"GAU\":\"D\", \"GAC\":\"D\", \"GAA\":\"E\", \"GAG\":\"E\",\n",
    "    \"GGU\":\"G\", \"GGC\":\"G\", \"GGA\":\"G\", \"GGG\":\"G\",}\n",
    "    \n",
    "    def __init__(self, seqstring):\n",
    "        super(RNASequence, self).__init__(seqstring)\n",
    "        return\n",
    "    \n",
    "    def transcribe(self):\n",
    "        return seq\n",
    "        \n",
    "class ProteinSequence(BaseSequence):\n",
    "    ValidChars = 'AMCEDGFIHKSTOPLNQPSRTWVY' #print \"\".join(list(set(pmap.values())))\n",
    "    \n",
    "    def __init__(self, seqstring):\n",
    "        super(ProteinSequence, self).__init__(seqstring)\n",
    "        return\n",
    "        \n",
    "    def load_PDB(self):\n",
    "        return\n",
    "\n",
    "class NGSCollection(object):\n",
    "    \n",
    "    def __init__(self, fastqfn):\n",
    "        \"\"\"Read from a fastq file into an appropriate Python native collection\"\"\"\n",
    "        return\n",
    "    \n",
    "    def QC_metrics(self):\n",
    "        \"\"\"Perform a few QC metrics and plots\"\"\"\n",
    "\n",
    "#bs = BaseSequence('kajsdfhslkjfhiugdfj')\n",
    "#dna = DNASequence('aCtgHIccctga')\n",
    "\n",
    "#classobj vs type: old style and new style classes\n",
    "#print type(BaseSequence)\n",
    "#print type(DNASequence)\n",
    "\n",
    "seq = \"\"\"\n",
    "ACAAGATGCCATTGTCCCCCGGCCTCCTGCTGCTGCTGCTCTCCGGGGCCACGGCCACCGCTGCCCTGCC\n",
    "CCTGGAGGGTGGCCCCACCGGCCGAGACAGCGAGCATATGCAGGAAGCGGCAGGAATAAGGAAAAGCAGC\n",
    "CTCCTGACTTTCCTCGCTTGGTGGTTTGAGTGGACCTCCCAGGCCAGTGCCGGGCCCCTCATAGGAGAGG\n",
    "AAGCTCGGGAGGTGGCCAGGCGGCAGGAAGGCGCACCCCCCCAGCAATCCGCGCGCCGGGACAGAATGCC\n",
    "CTGCAGGAACTTCTTCTGGAAGACCTTCTCCTCCTGCAAATAAAACCTCACCCATGAATGCTCACGCAAG\n",
    "TTTAATTACAGACCTGAA\n",
    "\"\"\"\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
